### PR TITLE
[WIP] Remove singleton .tar container from GraphReader

### PR DESCRIPTION
**Pull request by danpat**
# Issue

**NOTE:** `GraphReader` is not thread-safe, and this PR cannot be merged until it is made so.  The `cache_` member variable needs to become thread-safe, perhaps via some `thread_local` magic.
---

The `GraphReader` class has a singleton for the `tar_extract_` object - this is primarily to avoid accidentally parsing/loading a large `.tar` file more than once from multiple locations in the code.

Unfortunately, singletons are notoriously hard to test - if you wanted two different `GraphReader` objects that pointed to different `tile_extract` tar files you couldn't do it.  Once initialized, the `tile_extract` member variable of the `GraphReader` was fixed.  There was also no way to pass a subclassed `GraphReader` in to many of the places it was needed.

This PR changes all that.  It makes a few changes:
  - All `std::shared_ptr<GraphReader>` are changed to `GraphReader &`
  - The singleton `GraphReader::get_extract_instance` function is removed
  - locations inside workers/etc where `new GraphReader` was called previously are replaced with function parameters

This is essentially a conversion from the [Singleton Pattern](https://en.wikipedia.org/wiki/Singleton_pattern) to the [Dependency Injection Pattern](https://en.wikipedia.org/wiki/Dependency_injection).  The primary advantage is that test cases now have control over the `GraphReader` being used.  A minor beneficial side-effect *minor* advantage that there is a little bit less pointer de-referencing being done, and the `GraphReader` is now stack allocated in many cases.

## Tasklist

 - [x] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Generally use squash merge to rebase and clean comments before merging
 - [ ] Update the [changelog](CHANGELOG.md)

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

File: src/baldr/graphreader.cc [hunk 1]
```diff
@@ -23,50 +22,37 @@ constexpr size_t AVERAGE_MM_TILE_SIZE = 1024;         // 1k
...
-        // map files to graph ids
-        for (auto& c : archive->contents) {
-          try {
-            auto id = GraphTile::GetTileId(c.first);
-            tiles[id] = std::make_pair(const_cast<char*>(c.second.first), c.second.second);
-          } catch (...) {
-            // skip files we dont understand
-          }
+GraphReader::tile_extract_t::tile_extract_t(const boost::property_tree::ptree& pt) {
+  std::cout << "Initizliaing a tile extract" << std::endl;
```
Comment (mandeepsandhu) [inline src/baldr/graphreader.cc]: Change to LOG_INFO/DEBUG instead?

Comment (danpat) [reply src/baldr/graphreader.cc]: It's just temporary while I track down all the usages - will be removed in the final version.  I should've marked this PR as a draft.

**Comment by kevinkreiser** [comment]
with @kkaefer working on RAII changes to graphreader and a factory pattern coming to it after that i suggest we close this PR. @danpat what do you think?

**Comment by kevinkreiser** [comment]
great news! not so long ago we were able to implement this in a different way by simply speeding up the loading of a tar. now the memory map is still shared but we dont use a singleton to cache the index of the tar because you can create your tar with an index embedded in it. this was done over here: https://github.com/valhalla/valhalla/pull/3281

**Pull request closed without merging**